backend/cfg/vectorize.ml

```diff
@@ -112,24 +221,44 @@ end = struct
 
   let init () : t = Instruction.Id.Tbl.create 100
 
-  let from_basic_block (block : Cfg.basic_block) ~(dependency_graph : t) =
+  let get_all_dependencies dependency_graph id =
+    let (node : Node.t) = Instruction.Id.Tbl.find dependency_graph id in
+    node.all_dependencies
+
+  let get_all_dependencies_of_arg dependency_graph id ~arg_i =
+    let (node : Node.t) = Instruction.Id.Tbl.find dependency_graph id in
+    match node.reg_nodes.(arg_i).direct_dependency with
+    | None -> Instruction.Id.Set.empty
+    | Some direct_dependency ->
+      get_all_dependencies dependency_graph direct_dependency
+      |> Instruction.Id.Set.add direct_dependency
+
+  let from_block (block : Cfg.basic_block) =
+    let dependency_graph = init () in
     let is_changed_in instruction reg =
       Array.exists (Reg.same reg) (Instruction.results instruction)
       || Array.exists (Reg.same reg) (Instruction.destroyed instruction)
     in
-    (* CR-soon tip: break it into 2 parts to find the instruction we want then
-       go up from there. (currently it loops from the end and changes the answer
-       back to None when it encounters the same instruction) *)
     let latest_change ~(current : Instruction.Id.t) (reg : Reg.t) =
-      DLL.fold_right block.body
-        ~f:(fun basic_instruction latest ->
-          let instruction = Instruction.Basic basic_instruction in
-          if Instruction.Id.equal current (Instruction.id instruction)
-          then None
-          else if Option.is_none latest && is_changed_in instruction reg
+      let starting_cell =
+        match
+          DLL.find_cell_opt block.body ~f:(fun instruction ->
+              Basic instruction |> Instruction.id
+              |> Instruction.Id.equal current)
+        with
+        | None -> DLL.last_cell block.body
```

Comment:
In this case, not, but I used a similar function to find dependencies in the entire cfg instead of just one basic block before and back then this case would happen when current instruction is in another basic block.
Should I keep this or change it to something like assert false? Thanks!

Reply:
It's better to have "assert false" here if it's not supposed to be reachable.